In [8]:
# %matplotlib widget

from __future__ import annotations


import matplotlib.colors as mcolors
from test_utilities import test, TestDict

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 9: Movie Theater ---</h2><p>You <span title="wheeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee">slide down</span> the <a href="https://en.wikipedia.org/wiki/Fireman%27s_pole">firepole</a> in the corner of the playground and land in the North Pole base movie theater!</p>
<p>The movie theater has a big tile floor with an interesting pattern. Elves here are redecorating the theater by switching out some of the square tiles in the big grid they form. Some of the tiles are <em>red</em>; the Elves would like to find the largest rectangle that uses red tiles for two of its opposite corners. They even have a list of where the red tiles are located in the grid (your puzzle input).</p>
<p>For example:</p>
<pre><code>7,1
11,1
11,7
9,7
9,5
2,5
2,3
7,3
</code></pre>
<p>Showing red tiles as <code>#</code> and other tiles as <code>.</code>, the above arrangement of red tiles would look like this:</p>
<pre><code>..............
.......#...#..
..............
..#....#......
..............
..#......#....
..............
.........#.#..
..............
</code></pre>
<p>You can choose any two red tiles as the opposite corners of your rectangle; your goal is to find the largest rectangle possible.</p>
<p>For example, you could make a rectangle (shown as <code>O</code>) with an area of <code>24</code> between <code>2,5</code> and <code>9,7</code>:</p>
<pre><code>..............
.......#...#..
..............
..#....#......
..............
..<em>O</em>OOOOOOO....
..OOOOOOOO....
..OOOOOOO<em>O</em>.#..
..............
</code></pre>
<p>Or, you could make a rectangle with area <code>35</code> between <code>7,1</code> and <code>11,7</code>:</p>
<pre><code>..............
.......<em>O</em>OOOO..
.......OOOOO..
..#....OOOOO..
.......OOOOO..
..#....OOOOO..
.......OOOOO..
.......OOOO<em>O</em>..
..............
</code></pre>
<p>You could even make a thin rectangle with an area of only <code>6</code> between <code>7,3</code> and <code>2,3</code>:</p>
<pre><code>..............
.......#...#..
..............
..<em>O</em>OOOO<em>O</em>......
..............
..#......#....
..............
.........#.#..
..............
</code></pre>
<p>Ultimately, the largest rectangle you can make in this example has area <code><em>50</em></code>. One way to do this is between <code>2,5</code> and <code>11,1</code>:</p>
<pre><code>..............
..OOOOOOOOO<em>O</em>..
..OOOOOOOOOO..
..OOOOOOOOOO..
..OOOOOOOOOO..
..<em>O</em>OOOOOOOOO..
..............
.........#.#..
..............
</code></pre>
<p>Using two red tiles as opposite corners, <em>what is the largest area of any rectangle you can make?</em></p>
</article>


In [9]:
from itertools import batched, combinations
from math import prod
import re


tests: list[TestDict] = [
    {
        "name": "Example",
        "s": """
            7,1
            11,1
            11,7
            9,7
            9,5
            2,5
            2,3
            7,3
        """,
        "expected": 50,
    },
]
Tile = tuple[int, ...]


def surface(t1: Tile, t2: Tile) -> int:
    return prod(abs(d1 - d2) + 1 for d1, d2 in zip(t1, t2))


@test(tests=tests)
def part_I(s: str) -> int:
    tiles = sorted(batched(map(int, re.findall(r"\d+", s)), 2))
    return max(surface(t1, t2) for t1, t2 in combinations(tiles, 2))


Test Example passed, for part_I.
Success


In [10]:
with open("../input/day9.txt") as f:
    puzzle = f.read()

print(f"Part I: {part_I(puzzle)}")

Part I: 4760959496


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>4760959496</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>


<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The Elves just remembered: they can only switch out tiles that are <em>red</em> or <em>green</em>. So, your rectangle can only include red or green tiles.</p>
<p>In your list, every red tile is connected to the red tile before and after it by a straight line of <em>green tiles</em>. The list wraps, so the first red tile is also connected to the last red tile. Tiles that are adjacent in your list will always be on either the same row or the same column.</p>
<p>Using the same example as before, the tiles marked <code>X</code> would be green:</p>
<pre><code>..............
.......#XXX#..
.......X...X..
..#XXXX#...X..
..X........X..
..#XXXXXX#.X..
.........X.X..
.........#X#..
..............
</code></pre>
<p>In addition, all of the tiles <em>inside</em> this loop of red and green tiles are <em>also</em> green. So, in this example, these are the green tiles:</p>
<pre><code>..............
.......#XXX#..
.......XXXXX..
..#XXXX#XXXX..
..XXXXXXXXXX..
..#XXXXXX#XX..
.........XXX..
.........#X#..
..............
</code></pre>
<p>The remaining tiles are never red nor green.</p>
<p>The rectangle you choose still must have red tiles in opposite corners, but any other tiles it includes must now be red or green. This significantly limits your options.</p>
<p>For example, you could make a rectangle out of red and green tiles with an area of <code>15</code> between <code>7,3</code> and <code>11,1</code>:</p>
<pre><code>..............
.......OOOO<em>O</em>..
.......OOOOO..
..#XXXX<em>O</em>OOOO..
..XXXXXXXXXX..
..#XXXXXX#XX..
.........XXX..
.........#X#..
..............
</code></pre>
<p>Or, you could make a thin rectangle with an area of <code>3</code> between <code>9,7</code> and <code>9,5</code>:</p>
<pre><code>..............
.......#XXX#..
.......XXXXX..
..#XXXX#XXXX..
..XXXXXXXXXX..
..#XXXXXX<em>O</em>XX..
.........OXX..
.........<em>O</em>X#..
..............
</code></pre>
<p>The largest rectangle you can make in this example using only red and green tiles has area <code><em>24</em></code>. One way to do this is between <code>9,5</code> and <code>2,3</code>:</p>
<pre><code>..............
.......#XXX#..
.......XXXXX..
..<em>O</em>OOOOOOOXX..
..OOOOOOOOXX..
..OOOOOOO<em>O</em>XX..
.........XXX..
.........#X#..
..............
</code></pre>
<p>Using two red tiles as opposite corners, <em>what is the largest area of any rectangle you can make using only red and green tiles?</em></p>
</article>


In [14]:
from bisect import bisect, bisect_left
from collections.abc import Sequence
from dataclasses import dataclass
from itertools import batched, combinations, groupby, islice
from math import prod
import re
from typing import Self

from matplotlib import pyplot as plt
from more_itertools import first, flatten, last, minmax
import numpy as np
from util import Str

tests: list[TestDict] = [
    {
        "name": "Example",
        "s": """
            7,1
            11,1
            11,7
            9,7
            9,5
            2,5
            2,3
            7,3
        """,
        "expected": 24,
    },
]
Tile = tuple[int, int]
Edge = tuple[Tile, Tile]
WeightedInterval = tuple[int, int, int]


@dataclass
class Tree:
    x_mid: int  # split value
    center_left_sorted: list[
        WeightedInterval
    ]  # center intervals sorted by left endpoint ascending
    center_right_sorted: list[
        WeightedInterval
    ]  # center intervals sorted by right endpoint descending
    left_child: Tree | None  # Node or null
    right_child: Tree | None  # Node or null

    def stab_query(self, p: int) -> list[int]:
        output = []
        self._stab_query(p, output)
        return output

    def _stab_query(self, p: int, output: list[int]) -> None:
        if p < self.x_mid:
            #  scan center intervals sorted by left endpoint ascending
            for left, right, w in self.center_left_sorted:
                if left > p:
                    break
                if left <= p and p <= right:
                    output.append(w)

            if self.left_child:
                # recurse left
                self.left_child._stab_query(p, output)
            return

        if p > self.x_mid:
            # scan center intervals sorted by right endpoint descending
            for left, right, w in self.center_right_sorted:
                if right < p:
                    break
                if left <= p and p <= right:
                    output.append(w)

            if self.right_child:
                #  recurse right
                self.right_child._stab_query(p, output)
            return

        #  p == x_mid
        #  all center intervals contain p
        for left, right, w in self.center_left_sorted:
            output.append(w)

        # recurse both sides
        if self.left_child:
            self.left_child._stab_query(p, output)
        if self.right_child:
            self.right_child._stab_query(p, output)

    @classmethod
    def build(cls, intervals: Sequence[WeightedInterval]) -> Self | None:
        if not intervals:
            return None

        #  Step 1: collect all endpoints
        endpoints = sorted(flatten((l, r) for l, r, _ in intervals))

        #  Step 2: choose median split value
        mid_index = len(endpoints) // 2
        x_mid = endpoints[mid_index]

        #  Step 3: partition intervals
        center = []
        left_set = []
        right_set = []

        for left, right, w in intervals:
            if left <= x_mid <= right:
                center.append((left, right, w))
            elif right < x_mid:
                left_set.append((left, right, w))
            else:
                right_set.append((left, right, w))

        return cls(
            x_mid=x_mid,
            center_left_sorted=sorted(center, key=lambda t: t[:-1]),
            center_right_sorted=sorted(center, key=lambda t: t[1::-1], reverse=True),
            left_child=cls.build(left_set),
            right_child=cls.build(right_set),
        )


class Poly(Str):
    def __init__(self, s: str) -> None:
        tiles = [Tile(t) for t in batched(map(int, re.findall(r"\d+", s)), 2)]

        horizontal_edges = [
            WeightedInterval((*sorted(x for x, _ in ps), y))
            for y, ps in groupby(sorted(tiles, key=lambda t: t[::-1]), key=last)
        ]

        horizontal = Tree.build(horizontal_edges)

        verticall_edges = [
            WeightedInterval((*sorted(y for _, y in ps), x))
            for x, ps in groupby(sorted(tiles), key=first)
        ]

        vertical = Tree.build(verticall_edges)

        assert horizontal and vertical

        self.tiles = tiles
        self.horiontal = horizontal
        self.vertical = vertical

    def _valid_point_dimension(self, t1: Tile, t2: Tile, tree: Tree) -> bool:
        if t2 < t1:
            t1, t2 = t2, t1

        d11, d21 = t1
        d12, d22 = t2

        ranges1 = tree.stab_query(d21)
        ranges1.sort()

        i1, i2 = bisect(ranges1, d11), bisect_left(ranges1, d12)
        if i1 % 2 == 0 or i2 == len(ranges1):
            return False

        if i2 - i1 > 1:
            return False

        ranges2 = tree.stab_query(d22)
        ranges2.sort()

        i1, i2 = bisect(ranges2, d11), bisect_left(ranges2, d12)
        if i1 % 2 == 0 or i2 == len(ranges2):
            return False

        if i2 - i1 > 1:
            return False

        return True

    def valid_rectangls(self, t1: Tile, t2: Tile) -> bool:
        return self._valid_point_dimension(
            Tile(t1[::-1]), Tile(t2[::-1]), self.horiontal
        ) and self._valid_point_dimension(t1, t2, self.vertical)

    def largest_rectangle(self) -> int:
        max_surface = 0
        for t1, t2 in combinations(self.tiles, 2):
            if self.valid_rectangls(t1, t2):
                surf = surface(t1, t2)
                if surf > max_surface:
                    max_surface = surf

        return max_surface

    def plot(self) -> None:
        array = np.array(self.tiles + self.tiles[:1])
        x, y = array.T

        f = plt.figure(figsize=(5, 5), dpi=100)
        ax = f.add_axes([0, 0, 1, 1])  # type: ignore
        plt.plot(x, y, "-")
        plt.show()


@test(tests=tests)
def part_II(s: str) -> int:
    p = Poly(s)
    return p.largest_rectangle()


Test Example passed, for part_II.
Success


<link href="style.css" rel="stylesheet"></link>


In [15]:
print(f"Part II: {part_II(puzzle)}")

Part II: 1343576598


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>1343576598</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>


In [13]:
# intervals = [
#     (1, 4, 1),
#     (2, 6, 2),
#     (5, 8, 3),
#     (7, 9, 4),
#     (10, 12, 5),
#     (11, 15, 6),
#     (14, 18, 7),
# ]

# tree = Tree.build(intervals)
# assert tree is not None
# print(tree)
# sorted(w for _, _, w in tree.stab_query(2))